support-frontend/app/admin/settings/Switches.scala

```diff
@@ -1,21 +1,121 @@
+
+sealed trait SwitchState {
+  def isOn: Boolean
+}
+case object On extends SwitchState {
+  def isOn: Boolean = true
+}
+case object Off extends SwitchState {
+  def isOn: Boolean = false
+}
+
+object SwitchState {
+  import io.circe.generic.extras.semiauto.{ deriveEnumerationDecoder, deriveEnumerationEncoder }
+  implicit val stateDecoder: Decoder[SwitchState] = deriveEnumerationDecoder[SwitchState]
+  implicit val stateEncoder: Encoder[SwitchState] = deriveEnumerationEncoder[SwitchState]
+}
+
+case class FeatureSwitches(
+  enableQuantumMetric: SwitchState = Off,
+  usStripeAccountForSingle: SwitchState = On
+)
+
+case class CampaignSwitches(
+  enableContributionsCampaign: SwitchState = Off,
+  forceContributionsCampaign: SwitchState = Off
+)
+
+case class SubscriptionsSwitches(
+  enableDigitalSubGifting: SwitchState = On,
+  useDotcomContactPage: SwitchState = On,
+  checkoutPostcodeLookup: SwitchState = On
+)
+
+case class RecaptchaSwitches(
+  enableRecaptchaBackend: SwitchState = On,
+  enableRecaptchaFrontend: SwitchState = On
+)
+
+case class OneOffPaymentMethodSwitches(
+  stripe: SwitchState = On,
+  stripeApplePay: SwitchState = On,
+  stripePaymentRequestButton: SwitchState = On,
+  payPal: SwitchState = On,
+  amazonPay: SwitchState = On
+)
+
+case class RecurringPaymentMethodSwitches(
+  stripe: SwitchState = On,
+  stripeApplePay: SwitchState = On,
+  stripePaymentRequestButton: SwitchState = On,
+  payPal: SwitchState = On,
+  directDebit: SwitchState = On,
+  existingCard: SwitchState = On,
+  existingDirectDebit: SwitchState = On,
+  amazonPay: SwitchState = Off,
+  sepa: SwitchState = Off
+)
 
 case class Switches(
-  oneOffPaymentMethods: PaymentMethodsSwitch,
-  recurringPaymentMethods: PaymentMethodsSwitch,
-  enableDigitalSubGifting: SwitchState,
-  useDotcomContactPage: Option[SwitchState],
-  enableRecaptchaBackend: SwitchState,
-  enableRecaptchaFrontend: SwitchState,
-  experiments: Map[String, ExperimentSwitch],
-  enableContributionsCampaign: SwitchState,
-  forceContributionsCampaign: SwitchState,
-  enableQuantumMetric: SwitchState
+  oneOffPaymentMethods: OneOffPaymentMethodSwitches,
+  recurringPaymentMethods: RecurringPaymentMethodSwitches,
+  subscriptionsSwitches: SubscriptionsSwitches,
+  featureSwitches: FeatureSwitches,
+  campaignSwitches: CampaignSwitches,
+  recaptchaSwitches: RecaptchaSwitches
 )
 
 object Switches {
-  implicit val switchesCodec: Codec[Switches] = deriveCodec
+  /**
+   * Transforms from e.g:
+   *  {
+   *    "oneOffPaymentMethods" : {
+   *      "description" : "Payment methods - one-off contributions",
+   *      "switches" : {
+   *        "stripe" : {
+   *          "description" : "Stripe - Credit/Debit card",
+   *          "state" : "On"
+   *        }
+   *      }
+   *    }
+   *  }
+   *  To:
+   *  {
+   *    "oneOffPaymentMethods" : {
+   *      "stripe": "On"
+   *    }
+   *  }
+   */
+  private def flattenSwitchState(json: Json): Json =
+    root.state.string
```

Comment:
see https://circe.github.io/circe/optics.html if this is unfamiliar

Reply:
that is both neat and scary at the same time, I didn't know you could call a dynamic method name in scala!

Reply:
It's a little [scary](https://github.com/scala/scala/blob/v2.10.1/src/library/scala/Dynamic.scala#L23).
But - here it's not really any less safe than doing it the standard way, which can be a bit fiddly.
If `state` doesn't exist in the json then it quietly leaves it unchanged